Demo on measurement-based quantum computation

soosub opened the pull request:
### Before submitting

Please complete the following checklist when submitting a PR:

- [x] Ensure that your tutorial executes correctly, and conforms to the
      guidelines specified in the [README](../README.md).

- [x] Add a thumbnail link to your tutorial in `beginner.rst`, or if a
      QML implementation, in `implementations.rst`.

- [x] All QML tutorials conform to
      [PEP8 standards](https://www.python.org/dev/peps/pep-0008/).
      To auto format files, simply `pip install black`, and then
      run `black -l 100 path/to/file.py`.

------------------------------------------------------------------------------------------------------------

**Title:** Measurement-based quantum computation

**Summary:** A demo on measurement-based quantum computation

**Relevant references:**
- https://journals.aps.org/prl/abstract/10.1103/PhysRevLett.86.5188
- https://quantum-journal.org/papers/q-2021-02-04-392/

**Possible Drawbacks:**
None

**Related GitHub Issues:**
None

soosub commented:
Resources:
- https://phas.ubc.ca/~raussen/Supplements/VideoLectures.html
- https://arxiv.org/pdf/1711.04921.pdf

soosub commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,277 @@
+
+##############################################################################
+#
+# .. raw:: html
+#    :file: ../demonstrations/mbqc/rhg-graph.html
+#
```
Ideally I automatically generate the html instead of uploading it separately

soosub replied:
@josh146 do you happen to know if this is possible? It seems like the plotly output is not shown 😢

soosub replied:
I guess this might be related: https://github.com/readthedocs/sphinx_rtd_theme/issues/788

soosub commented:
Can someone add @isschoch and @mauriceweber as reviewers here?

isschoch commented:
Cannot wait to learn some cool stuff about MBQC!

soosub commented on demos_quantum-computing.rst:
```diff
@@ -83,6 +83,13 @@ such as benchmarking and characterizing quantum processors.
     :description: :doc:`demos/tutorial_photonics`
     :tags: quantumcomputing
+
+.. customgalleryitem::
+    :tooltip: Measurement-based quantum computation
+    :figure: demonstrations/mbqc/mbqc_blueprint.png
+    :description: :doc:`demos/tutorial_mbqc`
+    :tags: quantumcomputing
```
Somehow the gallery looks weird here. I tried to resolve it but have not found a solution yet.

![image](https://user-images.githubusercontent.com/44366402/181094285-c7f29e21-fc2a-4015-999d-493ca28224e3.png)

soosub commented on requirements.txt:
```diff
@@ -46,3 +46,7 @@ pyyaml==5.4.1
 jaxlib==0.1.64
 neural-tangents==0.3.6
+flamingpy==0.9.0b0
+plotly==5.9.0
+ipython==7.4.0
+nbformat>=4.2.0
```
Some of these might be removed when the demo is finalised

soosub commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,450 @@
+def one_bit_teleportation(input_state):
+    # Prepare input state
+    qml.QubitStateVector(input_state, wires=0)
+    qml.Hadamard(wires=1)
+
+    qml.CNOT(wires=[1, 0])
+    
+    # Measure the first qubit and apply an X-gate conditioned on the outcome
+    m = qml.measure(wires = [0])
+    qml.cond(m == 1, qml.PauliX)(wires=1)
```
Just a note that the circuit is not the same as the figure @radras

soosub commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,450 @@
+    return qml.state()
+
+
+print(qml.draw(cluster_state)())
```
To make the visualisation of the circuit more aesthetic, we can have a look at the comments in [this thread on slack](https://xanaduhq.slack.com/archives/C036VDB901W/p1653595712897369)

soosub replied:
@KetpuntoG do you have any suggestions for this? Is there a better way than using the `draw` method here?

soosub commented on demonstrations/tutorial_mbqc.py:
````````````````````````````diff
@@ -0,0 +1,461 @@
+# ```````````````````````````
+# Once we have established the ability to implement arbitrary single-qubit rotations and a two-qubit
+# gate, the CNOT, the final step is to show that we can implement arbitrary quantum circuits. To do
+# so, we simply have to note that we have a *universal gate set* [#DiVincenzo]_. However, you might wonder - how
````````````````````````````
I want to reference the DiVincenzo criteria paper here, still need to do that

soosub commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,461 @@
+#
+# Xanadu's approach toward a universal quantum computer involves continuous-variable cluster states 
+# [#CV-MBQC]_. If you would like to learn more about the architecture, you can read our blueprint 
+# papers [#XanaduBlueprint]_ and [#XanaduPassiveArchitecture]_. In the meantime on the hardware side, 
+# efforts are made to develop the necessary technology. This includes the recent `Borealis 
+# experiment <https://xanadu.ai/blog/beating-classical-computers-with-Borealis>`_ [#Borealis]_ where a 
+# 3-dimensional photonic graph state was created that was used to demonstrate quantum advantage. 
```
We can show the Borealis graph state here with FlamingPy and mention that it is not a cluster state.

soosub replied:
There are some intricate differences between CV/DV lands that would make it too complicated for this demo so let's not do it.

soosub commented on demonstrations/tutorial_mbqc.py:
````````````````````````diff
@@ -0,0 +1,461 @@
+#
+# .. _single-qubit-rotations:
+#
+# Single-qubit rotations
+# ```````````````````````
````````````````````````
In this section, we still want to add PennyLane snippets that show the equivalence between MBQC and the gate-based approach

soosub commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,461 @@
+# ----------------------
+# How do we know if this measurement-based scheme is just as powerful as its gate-based brother? We
+# have to prove it! In particular, we want to show that a measurement-based quantum computer is a
+# called quantum Turing machine (WTM) also known as a universal quantum computer. To do this, we
```
```suggestion
# called quantum Turing machine (QTM) also known as a universal quantum computer. To do this, we
```

soosub commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,461 @@
+# and :math:`|+\rangle = \frac{1}{\sqrt{2}}\big(|0\rangle + |1\rangle\big)` is the :math:`+1`` 
+# eigenstate of the Pauli-:math:`X` operator.
+#
+# The distinction between a graph state and a cluster state is that the latter has the additional 
+# condition that the underlying graph has to be a lattice. Physically, this means that ...
+# ... require Pauli measurements ... fully disentangle bla bla.
+#
```
We need to explain the distinction between graph states and cluster states here, and why it is physically and QIT-ically different

soosub replied:
Here are some relevant references:

1. [Persistent Entanglement in Arrays of Interacting Particles](https://journals.aps.org/prl/pdf/10.1103/PhysRevLett.86.910)
2. [Entanglement in Graph States and its Applications](https://arxiv.org/abs/quant-ph/0602096)
3. Section on Cluster States by Hans Briegel in the _Compendium of Quantum Physics_

soosub commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,461 @@
+
+a, b = 5, 1 # dimensions of the graph (lattice)
+G = nx.grid_graph(dim=[a, b]) # there are a * b qubits
+nx.draw(G, node_color='black')
```
Preferably, I want to have the figure a bit smaller

soosub replied:
and the linear chain should be straight

soosub commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,463 @@
+#   \frac{|0\rangle - e^{i\theta_j}|1\rangle}{\sqrt{2}}\right\},
+#
+# where the angles :math:`\theta_j`. Explicitly, these angles are given as
+#
+# .. :math:: \theta_{\mathrm{in}} = 0,  \theta_{1} = (-1)^{m_{\mathrm{in}} + 1} \alpha, 
+#   \theta_{2} = (-1)^{m_1} \beta, \theta_{3} = (-1)^{m_{\mathrm{in}} + m_2} \gamma
+# 
```
Somehow this does not render properly on the website

soosub commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,463 @@
+# .. math:: \hat{U}(\alpha, \beta, \gamma) = X^{m_1 + m_3}Z^{m_{\mathrm{in}} + m_2}U(\alpha, \beta, \gamma)
+#    |\psi_{\mathrm{in}}\rangle.
+#
+# Now note that this unitary :math:`\hat{U}` is related to our desired unitary :math:`U` up to 
+# possible the first two Pauli terms. However, we can correct for these additional Pauli gates by 
```
```suggestion
# Now note that this unitary :math:`\hat{U}` is related to our desired unitary :math:`U` up to 
# the first two Pauli terms. However, we can correct for these additional Pauli gates by 
```

soosub commented on demonstrations/tutorial_mbqc.py:
`````````````````````````diff
@@ -0,0 +1,463 @@
+one_bit_teleportation(input_state)
+
+##############################################################################
+# Propagation of logical information
+# ````````````````````````
+# Essentially, we keep logical information in one end of our cluster state which we progagate to the
+# other end using the teleportation protocol. By choosing adaptive measurements, we can "write" our
+# circuit onto the cluster state. Later, we will see how we can actually do this.
+#
+# It's good to emphasize that the entanglement of the cluster state is created *off-line*. This
+# means that the entanglement is made independently from the computation, like how a blank sheet of
+# paper is made separately from the text of a book. 
+#
`````````````````````````
This section is still incomplete

soosub commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,463 @@
+#
+
+##############################################################################
+#
```
I'd still like to explain dual and primal sheets, how it relates to the gate-based surface code etc.

soosub commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,512 @@
+# choosing the final the measurement basis appropriately or correcting for them classically after 
+# the quantum computation.
+#
+# Let's now see how this pans out in PennyLane. [WORK IN PROGRESS]
```
@radras can you take care of this?

soosub commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,512 @@
+#
+#    ..
+#
+# Let's now see how this pans out in PennyLane. [WORK IN PROGRESS]
```
and this too, please @radras?

mauriceweber reviewed (commented):
Very nice demo!! I learned quite a bit about MBQC 🎉 

I think the main point from my side is to make a concrete example for translating an arbitrary quantum circuit to a measurement based quantum computation (it's still a bit for me to imagine how exactly this is achieved). 

The quantum error correction section is a nice intro, although probably quite hard for people who don't now anything about QEC. So making a simple example with a small quantum circuit would illustrate MBQC for a wider audience.

soosub commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,512 @@
+# reliable way to produce qubits and stitch them together through entanglement, we can use it to
+# produce our cluster state resource! Essentially, we need some kind of qubit factory and a
+# stitching mechanism that puts it all together.
+#
```
Cool, I'll mention that this can be done with an Ising interaction or beamsplitters in case of photonics ⚡

soosub commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,512 @@
+# and :math:`|+\rangle = \frac{1}{\sqrt{2}}\big(|0\rangle + |1\rangle\big)` is the :math:`+1``
+# eigenstate of the Pauli-:math:`X` operator.
+#
+# The distinction between a graph state and a cluster state is that the latter has the additional
+# condition that the underlying graph has to be a lattice. Physically, this means that ...
+# ... require Pauli measurements ... fully disentangle bla bla.
```
This is still a TODO

soosub commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,687 @@
+# superconducting circuits.
+#
+#
+#   Cluster state entanglement + adaptive arbitrary single-qubit measurements = universal quantum computer
```
This box quote looks a bit weird rn, we can change this to make it look better

KetpuntoG commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,687 @@
+#
+# .. math:: |\psi_{\mathrm{out}}\rangle = \hat{U}(\alpha, \beta, \gamma)|\psi_{\mathrm{in}}\rangle 
+#    = X^{m_1 + m_3}Z^{m_{\mathrm{in}} + m_2}U(\alpha, \beta, \gamma)
+#    |\psi_{\mathrm{in}}\rangle.
```
I would have liked to follow the calculations, do you know any reference to put here to see it?

radras replied:
I added [this reference that I used](https://arxiv.org/pdf/2112.11601.pdf) although it doesn't have detailed derivations, I had to do some by hand to adapt the formulas for Rx and Rz.

KetpuntoG replied:
if it's not too complicated it would be great! 😄

soosub replied:
Could you include this reference @radras?

KetpuntoG reviewed (commented):
I leave you a few comments from the first part. On Monday I will finish the second half. I'm really enjoying it! Congratulations 😄
I think I understand the problem you commented @soosub . When talking about rotations you talk about 5 qubits but then you can't do more than one measurement, right? You could explain that those examples, the coefficients we know are zero and it is simplified somehow. But if not, it looks strange to see small circuits when you were expecting something with 5 qubits

soosub commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,687 @@
+    qml.Hadamard(wires=1)
+
+    # Return the density matrix of the output state
+    return qml.density_matrix(wires=[1])
```
Hmm in principle this should be a pure state, it's just that PennyLane does not support outputting the state vector if there was a mid-circuit measurement (which in general makes sense). Do you think we should still comment on this @KetpuntoG?

soosub replied:
Maybe we can show that the purity of the state / density matrix to get the point across

KetpuntoG replied:
by putting a link to wikipedia of what the density matrix is the first time it appears, it will be enough :)

soosub commented on demonstrations/tutorial_mbqc.py:
``````````````````````diff
@@ -0,0 +1,687 @@
+# One-bit Teleportation
+# `````````````````````
+# Let us have a deeper look at the principles behind the protocol using a simple example of one-bit
+# teleportation. We start with one qubit in the state :math:`|\psi\rangle` that we want to transfer
``````````````````````
Good call, we'll change that

KetpuntoG commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,687 @@
+# To demonstrate that this actually works, we will use PennyLane. For simplicity, we will just 
+# show the ability will to perform single-axis rotations :math:`R_z(\theta)` and 
+# :math:`R_x(\theta)`. Note these two plus the CNOT also constitute also a universal gate set. 
+# To start of, we define the :math:`R_z(\theta)` gate using two qubits with the gate-based approach.
```
remember to comment that being a case with 2 coefficients equal to 0, it is simplified and we do not need 5 qubits but 2. (same RX)

KetpuntoG commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,687 @@
+#    :align: center
+#    :width: 50%
+#
+#    ..
```
I missed the mathematical expression here. As you had done before in which expresses |phi_out> as a function of U and |phi_in>.

soosub replied:
@radras

radras commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,687 @@
+# .. figure:: ../demonstrations/mbqc/mbqc_info_flow.png
+#    :align: center
+#    :width: 75%
+#
```
If you look at the examples with a single-qubit rotation and a CNOT gate you can see how the state "flows" from left to right and ends up at the rightmost qubits. Looking at this figure you can relate any T shape with a 2-qubit gate and any straight path with a 1-qubit gate. Hence, in a sense, the layout of the MBQC protocol resembles the graphical representation of a circuit in a gate-based form where information flows through time while here information flows through space as well.

soosub replied:
I will put this in text and adjust the figure so that's clear later this week 🖌️

radras commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,687 @@
+# .. figure:: ../demonstrations/mbqc/one-bit-teleportation.png
+#    :align: center
+#    :width: 75%
+#
```
Well, this is just a convention we chose, maybe we should briefly mention the other possibility of applying a Hadamard. I think that this is more natural since this is what is actually happening. Take for example a case where you want to first perform a single-qubit unitary and then a CNOT. You can initialize all of the required qubits into the cluster state in the beginning. Then after a single qubit gate, you will have a qubit that will hold the rotated state |psi>. This state can then be considered as an input into a CNOT protocol and there is no need for another Hadamard since it has been performed at the beginning of the entire calculation to generate the cluster state. So basically I did this so it resembles the latter examples but we can change it if you think this might be confusing.

soosub replied:
We can put your comment above in the demo @radras 😏

soosub replied:
Can you include something that makes this clear in the text @radras?

KetpuntoG commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,687 @@
+# and :math:`y`) and one temporal dimension (:math:`z`). The cluster state alternates between *primal* and *dual sheets*, shown in the
+# figure above on the xy-plane. In principle, any quantum error correction code can be foliated into 
+# a graph state for measurement-based QEC. However, the foliations are particularly nice for CSS 
+# codes, named after Calderbank, Shor and Steane. CSS codes have stabilizers that exclusively contain
```
It would be great a link to deep into the CSS codes

soosub replied:
I don't understand what you mean @KetpuntoG

KetpuntoG replied:
Sorry, I mean to put this [link](https://en.wikipedia.org/wiki/CSS_code), or another one for CSS codes

soosub replied:
Ah I see, I will find a good reference. The wiki on CSS codes is shit 🙈

KetpuntoG reviewed (commented):
I have become a fan of MBQC! 🚀
I leave a couple more comments but I see it as very consistent

KetpuntoG commented on .github/workflows/build-dev.yml:
```diff
@@ -28,7 +28,7 @@ jobs:
         uses: actions/setup-python@v2
         with:
-          python-version: 3.7
+          python-version: 3.8
```
by the way @soosub @radras , what are all these changes you are making in these files. It seems like it's something you shouldn't be modifying

soosub replied:
I checked with @josh146 and if it doesn't break anything it's fine. We did it because FlamingPy requires Python 3.8.

KetpuntoG commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,732 @@
+# is the :math:`+1`` eigenstate of the Pauli-:math:`X` operator.
+#
+# The distinction between a graph state and a cluster state is that the latter has the additional
+# condition that the underlying graph has to be a lattice. Physically, this means that TODO ... 
```
I leave the comment so that this will not be forgotten

KetpuntoG commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,732 @@
+# Measurement-based quantum computation heavily relies on the idea of information propagation. In
+# particular, we make use of a protocol called *quantum teleportation*, one of the driving concepts behind MBQC. Despite its esoteric name, quantum
+# teleportation is very real and experimentally demonstrated multiple times in the last few decades 
+# [#Hermans2022]_, [#Furusawa1998]_, [#Riebe2004], [#Nielsen1998]. Moreover, it has related applications
```
Rendering failure, maybe the "_" has something to do with it?

soosub replied:
```suggestion
# [#Hermans2022]_, [#Furusawa1998]_, [#Riebe2004]_, [#Nielsen1998]_. Moreover, it has related applications
```

soosub replied:
It would be even nicer if we can group the references between one pair of brackets, do you know if that's possible @KetpuntoG ?

KetpuntoG commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,732 @@
+##############################################################################
+#
+# Seems good! As we can see, the resulting states are practically the same as we wanted to show. 
+# Note that we return `*density matrices* <https://en.wikipedia.org/wiki/Density_matrix>`_ in the 
```
I think you wanted to put the bold print here. On the other hand, the clarification of the density matrix I would put it in the teleportation part since it is the first time where it appears

soosub replied:
what line do you mean?

soosub replied:
@KetpuntoG do you know how to make the link text cursive?

KetpuntoG replied:
I've been looking at a few demos and haven't seen any in cursive. I don't know how to do it 🤔

josh146 replied:
Unfortunately, this is a limitation of reStructuredText. While it is generally much more powerful than markdown, unlike markdown it does not allow nested formatting!

There are [workarounds](https://stackoverflow.com/questions/4743845/format-text-in-a-link-in-restructuredtext) however, but they do feel a bit ugly.

soosub replied:
Thanks @josh146!

KetpuntoG commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,732 @@
+# About the authors
+# ----------------
+#
+# .. bio:: Joost Bus
```
The biographies have been decided to put in third person (I know that in your how-to it is in first person @soosub, but it will be changed)

soosub replied:
doesn't that sound terribly impersonal?

soosub replied:
I prefer first person

soosub replied:
*He prefers to speak in first person

KetpuntoG replied:
Ok, you can leave the bio in 1 person 👌

KetpuntoG reviewed (commented):
Completing the missing TODO's, for my part it's reviewed!
 As soon as it is ready, a second person will review and give the go-ahead 💪
 A very good demo topic, congratulations  🎉

KetpuntoG commented:
Hey! I cannot pass on the review to the next person until you complete the TODO's
Can you take a look at them? 🙂

soosub commented:
We're working on it @KetpuntoG 😌

soosub commented on requirements.txt:
```diff
@@ -46,3 +46,4 @@ pyyaml==5.4.1
 jax==0.3.15
 jaxlib==0.3.15
 neural-tangents==0.6.0
+flamingpy==0.9.0b0
```
this needs to be updated once FP PR#103 is in for the plotly plot

soosub commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,772 @@
+# choosing the measurement basis of qubit :math:`t_\mathrm{out}` appropriately or correcting for them classically after
+# the quantum computation.
+#
+# To demonstrate that this actually works, we will use PennyLane. For simplicity, we will just 
+# show the ability will to perform single-axis rotations :math:`R_z(\theta)` and 
+# :math:`R_x(\theta)` for arbitrary :math:`\theta \in [0, 2 \pi)`. Note that these two operations 
```
Hey @radras, could you add some more explanation of why we use less qubits here? Maybe also include the citation of the paper that you used

soosub commented:
Hey it's ready for another review. I still have two questions though:
- How do I put the floor operation in here? It seems like there are some problems in the surface code figure caption. Is this because it's a caption?
- Can I combine references into one pair of brackets? I have a couple of places where we reference multiple papers but it comes out as [1] [2] [3]. It would be nice to have something like [1, 2, 3] as per usual.

KetpuntoG commented:
> Hey it's ready for another review. I have one questions about formatting though:
> 
> * Can I combine references into one pair of brackets? I have a couple of places where we reference multiple papers but it comes out as [1] [2] [3]. It would be nice to have something like [1, 2, 3] as per usual.

Thanks! 
Not to my knowledge. It is something that appears in other demos.
 If I find out, I'll let you know 👌

soosub commented:
Ready for review @KetpuntoG

isaacdevlugt commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,809 @@
+
+dev = qml.device("default.qubit", wires=1)
+
+
```
```suggestion
```

isaacdevlugt commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,809 @@
+
+mbqc_dev = qml.device("default.qubit", wires=2)
+
+
```
```suggestion
```

isaacdevlugt commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,809 @@
+input_state = generate_random_state()
+theta = 2 * np.pi * np.random.random()
+
+np.allclose(RZ(theta, input_state), RZ_MBQC(theta, input_state))
```
🤯

isaacdevlugt commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,809 @@
+
+##############################################################################
+#
+# Finally, we again compare the two implementations with random state as an input.
```
```suggestion
# Finally, we again compare the two implementations with a random state as an input.
```

isaacdevlugt commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,809 @@
+#
+# Let's see how one can do this in PennyLane.
+
+# As sanity check, we will implement a CNOT gate on an arbitrary state.
+dev = qml.device("default.qubit", wires=2)
+
+
+@qml.qnode(dev)
+def CNOT(input_state):
+    # Prepare the input state
+    qml.QubitStateVector(input_state, wires=[0, 1])
+    qml.CNOT(wires=[0, 1])
+
+    return qml.density_matrix(wires=[0, 1])
+
+
+# Let's now implement a CNOT in MBQC formalism!
+# We will associate qubits c, t_in, a, and t_out in the figure with qubits 0, 1, 2, 3, respectively.
+mbqc_dev = qml.device("default.qubit", wires=4)
+
+
```
Just tidying up the code to reduce scroll time!

isaacdevlugt commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,809 @@
+# depends on the number of logical qubits, but also on the depth of the circuit. At this point,
+# it's good to reiterate that the entanglement of the cluster
+# state is created *off-line*. This means that the entanglement is made independently from the
+# computation, like how a blank sheet of paper is made separately from the text of a book.
```
Interesting analogy. I dig it 👍

isaacdevlugt commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,809 @@
+# it's good to reiterate that the entanglement of the cluster
+# state is created *off-line*. This means that the entanglement is made independently from the
+# computation, like how a blank sheet of paper is made separately from the text of a book.
+# Interestingly enough, we do not have to prepare all the entanglement at once. Just like we can
```
```suggestion
# Interestingly enough, we do not have to prepare all of the entanglement at once. Just like we can
```

isaacdevlugt commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,809 @@
+# computation, like how a blank sheet of paper is made separately from the text of a book.
+# Interestingly enough, we do not have to prepare all the entanglement at once. Just like we can
+# already start printing text upon the first few pages, we can apply measurements to one end of the
+# cluster, while growing it at the same time, as shown in the figure below. That is, we can start
```
```suggestion
# cluster while growing it at the same time, as shown in the figure below. That is, we can start
```

isaacdevlugt commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,809 @@
+# Interestingly enough, we do not have to prepare all the entanglement at once. Just like we can
+# already start printing text upon the first few pages, we can apply measurements to one end of the
+# cluster, while growing it at the same time, as shown in the figure below. That is, we can start
+# printing the text on the first few pages while at the same time reloading the printer's paper
+# tray!
```
You really went heavy into the printer paper analogy! 😆 I love it

isaacdevlugt commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,809 @@
+#
+#    ..
+#
+#    We can also consume the cluster state while we grow it. The blue qubits are in a cluster state,
+#    where the bonds between them represent entanglement. The gray qubits have been measured,
+#    destroying the entanglement and removing them from the cluster. At the same time, the green
+#    qubits are being added to the cluster by entangling them with it. Prior measurement outcomes
+#    determine the basis for future measurements [#OpticalQuantumComputing]_.
+#
```
This is rendering as the figure caption. Is that on purpose?

soosub replied:
Yes, do you think we should have a smaller caption and put the rest in the main text @isaacdevlugt?

isaacdevlugt replied:
No I think it's fine, but just make it a bit more explicit that it is indeed a figure caption. E.g.

```suggestion
#    A diagrammatic perspective on simultaneously consuming and growing a cluster state.
#    The blue qubits are in a cluster state,
#    where the bonds between them represent entanglement. The gray qubits have been measured,
#    destroying the entanglement and removing them from the cluster. At the same time, the green
#    qubits are being added to the cluster by entangling them with it. Prior measurement outcomes
#    determine the basis for future measurements [#OpticalQuantumComputing]_.
#
```

soosub replied:
👍🏼

isaacdevlugt commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,809 @@
+# reliable way to produce qubits and stitch them together through entanglement, we can use it to
+# produce our cluster state resource! Essentially, we need some kind of qubit factory and a
+# stitching mechanism that puts it all together. The stitching mechanism depends on the physical
+# platform, and can for example be implemented with an Ising interaction [#OneWay2001]_ or in the
+# context of photonics by interfering two modes with a beamsplitter.
```
```suggestion
# platform, and can, for example, be implemented with an Ising interaction [#OneWay2001]_ or, in the
# context of photonics, by interfering two modes with a beamsplitter.
```

isaacdevlugt commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,809 @@
+#
+# To mitigate the physical errors that can (and will) happen during a quantum computation, we
+# require some kind of error correction. Error correction is a technique for detecting errors and
+# reconstructing the logical data without losing any information. It is not exclusive to quantum computing;
+# it is also used in "classical" computing, memories and communication where one also has to deal `with
```
```suggestion
# reconstructing the logical data without losing any information. It is not exclusive to quantum computing;
# it is also used in "classical" computing, memories, and communication where one also has to deal `with
```

isaacdevlugt commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,809 @@
+# more errors in quantum computing, whereas the only errors in classical computing are bitflips; a 1
+# being flipped to a 0 or vice versa.
+#
+# A whole research field is devoted
+# to combating these challenges since Peter Shor published the seminal paper in 1995 [#ShorQEC1995]_.
+# Full coverage of this topic is beyond the scope of this tutorial, but a good place to start is
+# `Daniel Gottesman's thesis <https://arxiv.org/abs/quant-ph/9705052>`_ or `this blog post by
+# Arthur Pesah <https://arthurpesah.me/blog/2022-01-25-intro-qec-1/>`_ for a more compact
+# introduction. Instead, what we will do here is show how to implement error correction in the
+# MBQC framework by using the surface code [#FowlerSurfaceCode]_ [#FowlerPolyestimate]_ [#GoogleQEC2022]_ as
+# an example.
```
Excellent! :chefskiss:

isaacdevlugt commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,809 @@
+# MBQC framework by using the surface code [#FowlerSurfaceCode]_ [#FowlerPolyestimate]_ [#GoogleQEC2022]_ as
+# an example.
+#
+# .. figure:: ../demonstrations/mbqc/surface_code_d3.png
```
Nice!

isaacdevlugt commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,809 @@
+##############################################################################
+#
+# .. raw:: html
+#    :file: ../demonstrations/mbqc/rhg-graph.html
```
Oooooo! Pretty! 

My only concern is that the interactive picture here is _very_ large... Can you make it smaller by default?

soosub replied:
I can make it smaller 👍🏼

isaacdevlugt commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,809 @@
+# and :math:`y`) and one temporal dimension (:math:`z`). The cluster state alternates between *primal* and *dual sheets*, shown in the
+# figure above on the :math:`xy`-plane. In principle, any quantum error correction stabilizer code can be `foliated <https://arxiv.org/abs/1607.02579>`_ into
+# a graph state for measurement-based QEC [#FoliatedQuantumCodes]_, [#UniversalFTMBQC]_. However, the foliations are particularly nice for `CSS
+# codes <https://errorcorrectionzoo.org/c/css>`_, named after Calderbank, Shor and Steane. CSS codes have stabilizers that exclusively contain
```
```suggestion
# codes <https://errorcorrectionzoo.org/c/css>`_, named after Calderbank, Shor, and Steane. CSS codes have stabilizers that exclusively contain
```

isaacdevlugt commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,809 @@
+# outcomes of these measurements determine the measurement bases for future measurements, and the
+# last sheet of the lattice contains the encoded result of the computation which can be read out by yet another measurement.
+#
+# .. figure:: ../demonstrations/mbqc/gif_measuring.gif
```
🤩

isaacdevlugt commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,809 @@
+#
+
+##############################################################################
+# Quantum error correction
```
I think it needs to be made clear somewhere in this section that you won't be showing any code for error correction in the MBQC formalism. I got to the end of this section and was a little surprised. Obviously, it's totally fine that there's no code in this section! It's already a very long and amazing demo!

soosub replied:
Alright, what do you think should we put something like "giving the gist of QEC in the MBQC framework" in the section introduction? @isaacdevlugt

isaacdevlugt replied:
Yeah something like that! Maybe slightly less conversational. I wouldn't use "gist"

isaacdevlugt commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,809 @@
+#
+# .. container:: alert alert-block alert-info
+#
+#   **A system that is capable of creating an entangled cluster state and performing adaptive
+#   arbitrary single-qubit measurements on it is a universal quantum computer!**
```
Feels out of place 🤔

soosub replied:
Leave it out completely? Place it somewhere else? @isaacdevlugt

soosub replied:
I'll leave it out

isaacdevlugt commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,809 @@
+# video <https://www.youtube.com/watch?v=SD6TH7GZ1rM&feature=emb_title>`_ outlining the main ideas.
+# On the hardware side, efforts are made to develop the necessary technology. This includes the recent `Borealis
+# experiment <https://xanadu.ai/blog/beating-classical-computers-with-Borealis>`_ [#Borealis]_ where
+# a 3-dimensional photonic graph state was created that was used to demonstrate quantum advantage.
```
```suggestion
# a 3-dimensional photonic graph state was created that was used to demonstrate quantum computational advantage.
```

isaacdevlugt reviewed (changes requested):
Really great demo guys! Comments are all very minor. Nice work!

soosub commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,809 @@
+# [#Hermans2022]_, [#Furusawa1998]_, [#Riebe2004]_, [#Nielsen1998]_. Moreover, it has related applications
+# in safe communication protocols that are impossible with classical communication so it's certainly
+# worth learning about. In this protocol, we do not transport matter but *information* between systems. Admittedly, it has a
+# rather delusive name because it is not instantaneous but requires communication of additional classical information,
+# which is naturally limited by the speed of light.
```
Hmm what I tried to convey is that it's we don't have faster than light communication. Should we rephrase?

soosub commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,809 @@
+
+##############################################################################
+#
+# Note that we return a `density matrix <https://en.wikipedia.org/wiki/Density_matrix>`_ in the
+# function above because we are dealing with a *quantum channel* that includes a measurement on a
+# subset of the qubits. These matrices allow for description of mixed quantum states and are an
+# extension of pure state vectors. Using this formalism, we can describe operations beyond unitaries
+# such as the teleportation protocol.
+#
+# Now, let's prepare a random qubit state and see if the teleportation protocol is working as
```
Do you want to omit the reason why we use density matrices?

soosub replied:
That seems to leave away a lot of info, don't you think? @isaacdevlugt

isaacdevlugt replied:
Hmm yea we can find some middle ground here. How about this?

```suggestion
# Note that we return a `density matrix <https://en.wikipedia.org/wiki/Density_matrix>`_ in the
# function above.
# This allows for us to describe operations beyond unitaries
# such as the teleportation protocol.
#
# Now, let's prepare a random qubit state and see if the teleportation protocol is working as
```

soosub commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,809 @@
+input_state = generate_random_state()
+theta = 2 * np.pi * np.random.random()
+
+np.allclose(RX(theta, input_state), RX_MBQC(theta, input_state))
```
Shout out to @radras 🙌🏼

soosub commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,809 @@
+
+##############################################################################
+#
+# This is a fairly simple cluster state, but we `will later see <#single-qubit-rotations>`_ how even
```
```suggestion
# This is a fairly simple cluster state, but we will later see how even
```

soosub commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,809 @@
+# We then apply the teleportation protocol and see if the resulting density matrix of the output
+# state of the second qubit is the same as the input state of the first qubit.
+
+# Define helper function for random input state on n qubits
```
Hmm I prefer it this way

isaacdevlugt replied:
okay! I would maybe leave comments in the code block then. E.g., 

```python
def generate_random_state(n=1):
    input_state = np.random.random(2 ** n) + 1j * np.random.random(2 ** n) # psi = alpha |0> + beta |1>`
    return input_state / np.linalg.norm(input_state)
```

or something like that

soosub commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,809 @@
+#
+#    ..
+#
+# The input qubit :math:`t_\mathrm{in}`, together with the intermediate qubits :math:`a_1`,
```
That will get a bit messy 🤔 You would have to keep track of density matrices and properly do the projective measurements. I don't think showing it here will make thinks clearer.

isaacdevlugt replied:
Okay! Good to know. Was just wondering if it would be feasible.

soosub commented:
```
from flamingpy.codes import SurfaceCode
from plotly.subplots import make_subplots
import plotly.graph_objects as go

# RHG lattice
RHG = SurfaceCode(3)

# Layout
size = 450
fig = make_subplots(rows=1, cols=2, 
                    specs=[[{'type':'scene'}, {'type':'scene'}]], 
                    column_titles=["Primal", "Dual"], 
                    horizontal_spacing = 0.02)
fig.update_layout(width=2*size, height=size)

# Primal
x = 2
fig1 = RHG.draw(backend="plotly", label=["type"], showbackground=True, width=size, height=size)
fig.update_layout(scene = dict(xaxis = {'nticks': 2, 'range': [x+0.5, x-0.5], 'dtick':1}), showlegend=False)
fig.update_layout(scene = dict(yaxis = {'dtick':1}))
fig.update_layout(scene = dict(zaxis = {'dtick':1}))
fig.add_trace(fig1.data[0], row=1, col=1)
fig.add_trace(fig1.data[1], row=1, col=1)

# Dual
x = 1
fig2 = RHG.draw(backend="plotly", label=["type"], showbackground=True, width=size, height=size)
fig.get_subplot(1,2)['xaxis'].update({'nticks': 2, 'range': [x+0.5, x-0.5]})
fig.get_subplot(1,2)['yaxis'].update({'dtick':1})
fig.get_subplot(1,2)['zaxis'].update({'dtick':1})
fig.add_trace(fig2.data[0], row=1, col=2)
fig.add_trace(fig2.data[1], row=1, col=2)

# Showing figure
fig.show(width=size, height=size)

# Saving figure
fig.write_image("primal_dual.jpeg")
fig.write_html("primal_dual.html")
```
![primal_dual](https://user-images.githubusercontent.com/44366402/186234209-c12350cd-46b5-4d2c-8a5d-739adb0f783c.jpeg)

isaacdevlugt commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,809 @@
+# of earlier measurements on the chain. After these operations, the state of qubit
+# :math:`t_\mathrm{out}` is given by
+#
+# .. math:: |\psi_{\mathrm{out}}\rangle = \hat{U}(\alpha, \beta, \gamma)|\psi_{\mathrm{in}}\rangle
```
It's a small pet peeve of mine that we don't put hats on anything to begin with, but it's a settled-upon convention in the quantum community so it is what it is. But, either everything should have a hat on it, or no hats at all in my opinion. 

I'd opt for no hats _anywhere_, and then renaming `\hat{U}` to something else, like `V` for example.

soosub replied:
Ah I see the confusion. I am using the hat just to make clear that U and \hat{U} are different, not to designate that \hat{U} is a operator. I could use a tilde or apostrophe instead @isaacdevlugt

isaacdevlugt reviewed (changes requested):
Just addressing comments

soosub commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,812 @@
+
+code_distance = 3
+RHG = SurfaceCode(code_distance)
+# RHG.draw(backend="plotly", label=["type"], showbackground=True) #TODO: uncomment this line after merging FP#103
```
TODO ⚠️

soosub replied:
Installing it with git causes errors for some reason. We might need to wait for this month's release of FlamingPy before publishing.

soosub reviewed (commented):
Just adding a note here that the demo should not be posted before these changes are added to FlamingPy

soosub commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,818 @@
+# look at the :ref:`figure <fig-surfacecode>` with the distance-3 surface code and try to link it 
+# with the dual and primal sheets shown here!
+#
+# .. raw:: html
+#    :file: ../demonstrations/mbqc/primal_dual.html
+#
```
@ilantz I think I might have interchanged primal/dual. Am I correct?

ilan-tz replied:
Primal sheets are at even Z slices, so they are flipped in the figure

josh146 commented on requirements.txt:
```diff
@@ -46,3 +46,4 @@ pyyaml==5.4.1
 jax==0.3.15
 jaxlib==0.3.15
 neural-tangents==0.6.0
+git+https://github.com/XanaduAI/flamingpy@main
```
```suggestion
git+https://github.com/XanaduAI/flamingpy.git
```

ilan-tz commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,818 @@
+#    logical qubit and can correct any combination of :math:`\lfloor (d-1)/2 \rfloor` errors.
+#    [#FowlerPolyestimate]_
+#
+# In the measurement-based picture, quantum error correction requires cluster states that are at
+# least 3-dimensional [#XanaduBlueprint]_, contrary to the 2-dimensional cluster states required for
+# universal quantum computation discussed in the previous section. The error correcting code that you want to implement
```
```suggestion
# In the measurement-based picture, successive rounds of quantum error correction requires generating cluster states
# that are at least 3-dimensional [#XanaduBlueprint]_, contrary to the 2-dimensional cluster states required for
# universal quantum computation discussed in the previous section. The error correcting code that you want to implement
```

ilan-tz commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,818 @@
+# a graph state for measurement-based QEC [#FoliatedQuantumCodes]_, [#UniversalFTMBQC]_. However, the foliations are particularly nice for `CSS
+# codes <https://errorcorrectionzoo.org/c/css>`_, named after Calderbank, Shor, and Steane. CSS codes have stabilizers that exclusively contain
+# :math:`X`-stabilizers *or* :math:`Z`-stabilizers, and include the *surface code* and *colour code* families.
+# For these CSS codes, you can view the primal and dual sheets as measuring the
```
```suggestion
# For these CSS codes, you can roughly view the primal and dual sheets as measuring the
```

ilan-tz commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,818 @@
+# -------------------------------
+#
+# We have learned that a one-way quantum computer capable of cluster state
+# entanglement together with adaptive arbitrary single-qubit measurements allows for universal
```
```suggestion
# entanglement together with adaptive arbitrary single-qubit measurements allows for universal
```
Consider modifying in light of my comments above

soosub replied:
I think it makes sense to assume that ability here, and then later state that this is not a strict requirement. What do you think?

soosub replied:
https://github.com/PennyLaneAI/qml/pull/485/files#r955431398 nevermind 😌

soosub replied:
I mention the assumption now in the section on universality 👍🏼

ilan-tz commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,818 @@
+#    qubits are being added to the cluster by entangling them with it. Prior measurement outcomes
+#    determine the basis for future measurements [#OpticalQuantumComputing]_.
+#
+# This feature makes it particularly interesting for photonic-based quantum computers: we can use
+# expendable qubits that don't have to stick around for the full calculation. If we can find a
+# reliable way to produce qubits and stitch them together through entanglement, we can use it to
```
```suggestion
# This feature makes it particularly attractive for photonic quantum computers: we can use
# expendable qubits that don't have to stick around for the full calculation. If we can find a
# reliable way to produce qubits and stitch them together through entanglement, we can use it to
```
The don't stick around, but they also can't, because they are likely to be lost or corrupted by loss

ilan-tz commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,818 @@
+# We have learned that a one-way quantum computer capable of cluster state
+# entanglement together with adaptive arbitrary single-qubit measurements allows for universal
+# quantum computation. The MBQC framework is a powerful quantum computing approach, particularly
+# useful in platforms that allow for many expendable qubits, and offers
```
Consider: platforms with flying qubits and easy physical entangling gates

ilan-tz commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,818 @@
+# quantum computation. The MBQC framework is a powerful quantum computing approach, particularly
+# useful in platforms that allow for many expendable qubits, and offers
+# several advantages over the gate-based model. For example, it circumvents the need for applying
+# in-line entangling gates which are often the most noisy operations in gate-based quantum computers
+# with trapped-ions or superconducting circuits. Instead, the required entanglement is created
+# off-line which is often simpler to implement.
```
For photonics: the depth of the optical circuit should not increase with the depth of the logical circuit because (a) the losses become intolerable, and (b) for encodings with probabilistic gates, e.g. dual rail entangling gates, the success probability of the computation will exponentially decay

ilan-tz commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,818 @@
+# with trapped-ions or superconducting circuits. Instead, the required entanglement is created
+# off-line which is often simpler to implement.
+#
+# In this demo, we assumed that the system is capable of performing arbitrary
+# single-qubit measurements. This is not a strict requirement, as one can acquire the same
+# capabilities by sprinkling *magic states* into the cluster state. A discussion of this topic is
+# beyond the scope of this tutorial, but a good place to start is `this
+# paper <https://arxiv.org/abs/quant-ph/0403025>`_ [#MagicStates]_.
```
Ah, I see you address this. I think it may be worthwhile to insinuate this earlier on

ilan-tz commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,818 @@
+# video <https://www.youtube.com/watch?v=SD6TH7GZ1rM&feature=emb_title>`_ outlining the main ideas.
+# On the hardware side, efforts are made to develop the necessary technology. This includes the recent `Borealis
+# experiment <https://xanadu.ai/blog/beating-classical-computers-with-Borealis>`_ [#Borealis]_ where
+# a 3-dimensional photonic graph state was created that was used to demonstrate quantum computational advantage.
```
![image](https://user-images.githubusercontent.com/57886357/186771104-8cd3ac7f-0240-4cd7-953d-49bfd4623e84.png)

soosub replied:
What do you mean? Don't you think graph state is the right term? @ilan-tz

soosub replied:
Just had a talk with Raf: he agreed and I will just drop the connection with the Borealis experiment

soosub commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,818 @@
+# corresponding graph state.
+#
+
+import networkx as nx
+import matplotlib.pyplot as plt
+
+a, b = 1, 5  # dimensions of the graph (lattice)
+G = nx.grid_graph(dim=[a, b])  # there are a * b qubits
+
+plt.figure(figsize=(5, 1))
+nx.draw(G, pos={node: node for node in G}, node_size=500, node_color="black")
```
I think that might complicate things too much.

soosub commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,818 @@
+# the computation is not actually computing anything. In fact, aside from the width and depth of
+# the desired quantum circuit, the cluster state generation is essentially independent of the
+# calculation. If you have a reliable way of applying these two operations (Hadamard and
+# controlled-:math:`Z`), you are ready for the next step: worrying about conditional single-qubit
+# measurements.
```
first worry, then perform 😌

soosub commented on demonstrations/tutorial_mbqc.py:
```diff
@@ -0,0 +1,818 @@
+#
+#    ..
+#
+#    A distance-3 surface code. Circles represent qubits and bubbles represent operators, called
```
```suggestion
#    A distance :math:`d=3` surface code. Circles represent qubits and bubbles represent operators, called
```

soosub commented on requirements.txt:
```diff
@@ -46,3 +46,5 @@ pyyaml==5.4.1
 jax==0.3.15
 jaxlib==0.3.15
 neural-tangents==0.6.0
+git+https://github.com/XanaduAI/flamingpy@mbqc-demo
```
> **Warning**: this will fail once the branch is deleted. I opened up a [PR here](https://github.com/XanaduAI/flamingpy/pull/118) to get the changes from the `mbqc-demo` branch in flamingpy, but eventually we'd want to use the `pip` version anyway.

soosub replied:
Fixed https://github.com/XanaduAI/flamingpy/pull/118

soosub commented on demonstrations/tutorial_mbqc.py:
````````````````````````diff
@@ -0,0 +1,818 @@
+#
+# Single-qubit rotations
+# ```````````````````````
+# Arbitrary single-qubit rotations are an essential operation for a universal quantum computer. In
````````````````````````
imo that's just a different interpretation - they are just implemented in a different way (and still essential)

soosub commented on requirements.txt:
```diff
@@ -46,3 +46,5 @@ pyyaml==5.4.1
 jax==0.3.15
 jaxlib==0.3.15
 neural-tangents==0.6.0
+git+https://github.com/XanaduAI/flamingpy
+plotly>=4.5.0
```
Plotly can be uninstalled when FlamingPy is installed with `pip`. However, this should wait until this month's (August 2022) release of FlamingPy.

soosub commented:
@KetpuntoG @isaacdevlugt I addressed all the comments. There are little things that @radras might want to address still, but I don't think they are strictly necessary. We could always update the demo in the future. I will leave the finalisation in your capable hands 👍🏼

KetpuntoG commented:
Don't worry about updating the branch @soosub , the demo is perfect 😄 when we upload it we will update everything

soosub commented:
@KetpuntoG I was planning to address [this ](https://github.com/PennyLaneAI/qml/pull/485#discussion_r956315514) now the FlamingPy August release is out but there [seem to be some issues](https://xanadu-quantum.slack.com/archives/C03MEBR3T5E/p1662741868767199?thread_ts=1661910934.558379&cid=C03MEBR3T5E) 🥵

KetpuntoG commented:
> @KetpuntoG I was planning to address [this ](https://github.com/PennyLaneAI/qml/pull/485#discussion_r956315514) now the FlamingPy August release is out but there [seem to be some issues](https://xanadu-quantum.slack.com/archives/C03MEBR3T5E/p1662741868767199?thread_ts=1661910934.558379&cid=C03MEBR3T5E) 🥵

Ah okey! Let's see if Narinam answer you :)

isaacdevlugt reviewed (approved):
My comments were addressed or resolved!

KetpuntoG commented:
@soosub , @radras thanks for this great demo! 
Tomorrow we will announce the release of this demo on Pennylane twitter 🚀
